lib/Dialect/ESI/runtime/cpp/lib/backends/Cosim.cpp

```diff
@@ -415,9 +415,81 @@ class CosimMMIO : public MMIO {
   std::unique_ptr<FuncService::Function> cmdMMIO;
 };
 
+#pragma pack(push, 1)
+struct HostMemReadReq {
+  uint8_t tag;
+  uint32_t length;
+  uint64_t address;
+};
+
+struct HostMemReadResp {
+  uint64_t data;
+  uint8_t tag;
+};
+#pragma pack(pop)
+
 class CosimHostMem : public HostMem {
 public:
-  CosimHostMem() {}
+  CosimHostMem(AcceleratorConnection &acc, Context &ctxt,
+               StubContainer *rpcClient)
+      : acc(acc), ctxt(ctxt), rpcClient(rpcClient) {}
+
+  void start() override {
+    // We have to locate the channels ourselves since this service might be used
+    // to retrieve the manifest.
+
+    // Setup the read side callback.
+    ChannelDesc readArg, readResp;
+    if (!rpcClient->getChannelDesc("__cosim_hostmem_read.arg", readArg) ||
+        !rpcClient->getChannelDesc("__cosim_hostmem_read.result", readResp))
```

Comment:
Not the nicest to tightly couple this to `frontends/PyCDE/src/pycde/bsp/cosim.py` through a string constant. I understand that there isn't a neat way to share this string between C++ and Python code, but is this really the best way of doing it?

Reply:
I can't think of any given that the runtime and PyCDE are not technically coupled. I could imagine a way to keep those constants in the runtime and suck them into CIRCT and/or PyCDE at compile time... but it wouldn't be pretty.